internal/cluster/clusterworkflow/activity_create_node_pool.go

```diff
@@ -191,6 +195,15 @@ func (a CreateNodePoolActivity) Execute(ctx context.Context, input CreateNodePoo
 
 		_, err = eksworkflow.NewCreateAsgActivity(a.awsSessionFactory, nodePoolTemplate).Execute(ctx, subinput)
 		if err != nil {
+			input := eksworkflow.DeleteStackActivityInput{
```

Comment:
Did you consider handling this branch in the workflow, and deleting the stack in a different activity? It may be more robust, for example because of the retries and the delays between the creation and the deletion requests (and also the result of the workflow is easier to understand in case of failures). Otherwise, please annotate the stack delete errors returned with the intention (like wrap with "deleting failed stack")

Reply:
I was thinking about it. I could argue for and against both, whichever you would prefer is fine for me.

I guess I dropped it in favor of my original solution because on the level of the workflow if the `CreateNodePoolActivity` activity fails then we **MUST** execute the inverse operation of each mutating subactivity potentially executed in the mentioned main activity - or identify the error and branch from there to undo the previous actually executed mutating subactivities.
Which would loosely couple the activity and workflow code/logic which is IMO a fragile approach from maintenance perspective because of the distance inbetween/scope encapsulating them.
It was a higher precedence for me to avoid this - details on this can be requested.

Also thought about either modifying the `DeleteNodePoolActivity` or defining an `UndoCreateNodePoolActivity` activity which is basically a delete activity with try/no-fast-fail subactivities (so errors doesn't immediately terminate the function) but it would have required more time to think about so I wanted to start the discussion by opening the PR with my original approach.
It has its ups and downs, I would somewhat like that approach if we can pay attention to it on a repository scale and handle activities accordingly from a point in time - which could not be fulfilled on my last project unless I personally made sure of it so I'm a bit more careful this time.
It also means that the create and delete[+undo] activities are needed to be kept "in sync" as each other's inverse operation, similarly the workflow+activity combination discussed earlier.

I guess I wanted to narrow the scope down as much as possible.

On the other hand my approach can also end up badly if another mutating subactivity is introduced and the error handling logic is not updated - but I guess a function's scope is still better than a package's/workflow-activity pair's.
I'm going to think about better ways to do this[1], I might have hastily decided on the first approach, but as a discussion driver I achieved my goal.
[1]: one quick idea is to define a local cleanup function which can work on such a try/no-fast-fail basis, and wire the error handling through that, it seems to have the ups and minimize the downs, but this is just a McDonalds idea popping into my head while I were typing the response.

Reply:
I will definitely incorporate the error changes you mentioned, adding it to the todo list